### DISCUSSION TITLE
Best way to add additional data in public.users table

### DISCUSSION THREAD

**Discussion by inbytesdevteam** [supabase/supabase#556] [Questions]
I'd like to be able to get a user's first name, last name, and username as they sign up. In the [guide](https://supabase.io/docs/guides/auth#create-a-publicusers-table), it says that I can create a `public.users` table with the `uuid`.

Is it possible to add in the details above when using `auth.signUp()`? Or should I use `auth.update()` to update after sign up? I'd like for it to show an error if the username is already taken. The problem I find with having two commands is I may have to delete the entry in the `auth.users` table if adding it to `public.users` fails. I'm not sure if I'm missing anything. What's the best way to go about this?

**Comment by kiwicopple** [comment]
For now the recommended way is to to an `update` after signup - as you said. See an example here:
https://github.com/vercel/nextjs-subscription-payments/blob/main/pages/signup.js

> I may have to delete the entry in the auth.users table if adding it to public.users fails. 

Make sure you use a trigger to insert the new user! Like this:
https://github.com/vercel/nextjs-subscription-payments/blob/96eb86968f91a6e99273bec86894fd3f0103fac1/schema.sql#L22

**Reply by inbytesdevteam** [reply]
Thanks for this @kiwicopple! I tried the example above but it seems like the trigger returns a 500 error:

```sql
create function public.handle_new_user() 
returns trigger as $$
begin
  insert into public.users (id, full_name, avatar_url)
  values (new.id, new.raw_user_meta_data->>'full_name', new.raw_user_meta_data->>'avatar_url');
  return new;
end;
```

Since this trigger is done after `signUp` and **before** `update` of users, where do you add the `raw_user_meta_data` then?
```js
const { error, user } = await signUp({ email, password });
  if (error) {
    setMessage({ type: 'error', content: error.message });
  } else {
  if (user) {
    await supabase.from('users').update({ full_name: name }).eq('id', user.id);
    setUser(user);
  } else {
    setMessage({
      type: 'note',
      content: 'Check your email for the confirmation link.'
    });
  }
}
```

**Reply by kiwicopple** [reply]
I think it might be better to do:

```sql
create function public.handle_new_user() 
returns trigger as $$
begin
  insert into public.users (id)
  values (new.id);
  return new;
end;
```

Then you can insert all your user data directly into the public table? I'm guessing there is a problem using the `new.raw_user_meta_data` value using your auth flow but I would need to see a code example (for example, you have a different setting for email confirmations than the example above?)

Note there is a one-click deploy button for that example:
<img width="206" alt="image" src="https://user-images.githubusercontent.com/10214025/105285322-215d3500-5bef-11eb-99bb-c26c2b232fe8.png">

You can get it working in ~1 minute to compare your settings
